apollo-gradle-plugin/src/main/kotlin/com/apollographql/apollo/gradle/api/ApolloGenerateOperationIdsTask.kt

```diff
@@ -0,0 +1,48 @@
+package com.apollographql.apollo.gradle.api
+
+import com.apollographql.apollo.compiler.operationoutput.OperationList
+import com.apollographql.apollo.compiler.operationoutput.OperationOutput
+import com.apollographql.apollo.compiler.operationoutput.toJson
+import org.gradle.api.DefaultTask
+import org.gradle.api.file.FileCollection
+import org.gradle.api.file.RegularFileProperty
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.OutputFile
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+import org.gradle.api.tasks.SkipWhenEmpty
+import org.gradle.api.tasks.TaskAction
+
+@CacheableTask
+abstract class ApolloGenerateOperationIdsTask : DefaultTask() {
```

Comment:
Does this class actually need to be extended by users? Couldn't it work with just an interface containing `generateOperationOutput`, which gets invoked by a task managed by the Apollo plugin?

Reply:
I thought you preferred defining your own Task. Having an interface, like `OperationIdGenerator` is of course doable too. It just feels a bit less "Gradle-y" because users will have to remember to bump `operationIdGenerator.version` every time the implementation changes.

With a user-defined task, Gradle will magically determine the classpath and invalidate when needed.

Reply:
The classpath invalidation is a good point.

Reply:
At this stage, I think you'll be the first user of this API so I'm happy to modify OperationIdgenerator with something like below instead:

```kotlin
abstract class OperationIdGenerator {
  abstract fun apply(operationDocument: String, name: String, packageName: String)

  // overridee for batch handling
  open fun generateOperationOutput(operationList: OperationList): OperationOutput {
    operationList.map {
      apply(it.source, it.name, it.packageName) to it
    }
  }

  // change version here everytime the implementation of this class changes
  val version: String
}
```

Reply:
Or if we want an interface:

```kotlin
interface BatchOperationIdGenerator {
  fun generateOperationOutput(operationList: OperationList): OperationOutput

  // change version here everytime the implementation of this class changes
  val version: String
}
```

Reply:
I do think the automatic classpath invalidation is worth extending the custom task. Less mental load.

Reply:
👍 sounds good